### ISSUE TITLE
API is unsound. Segfault and double free is possible.

### ISSUE THREAD

**Issue by steffahn** [jonhoo/rust-evmap#78]
```rust
use std::sync::atomic::{AtomicBool, Ordering::SeqCst};
static SNEAKY: AtomicBool = AtomicBool::new(false);

#[derive(Eq, Hash)]
struct Sneaky(Vec<u8>);
impl PartialEq for Sneaky {
    fn eq(&self, other: &Self) -> bool {
        if SNEAKY.swap(false, SeqCst) {
            false
        } else {
            self.0 == other.0
        }
    }
}

fn main() {
    let (_, mut w) = evmap::new::<u8, Box<Sneaky>>();
    w.insert(0, Box::new(Sneaky(vec![0])));
    w.refresh();
    w.remove(0, Box::new(Sneaky(vec![0])));
    SNEAKY.store(true, SeqCst);
    w.refresh();
    w.refresh();
    w.refresh();
}
```
```rust
cargo run
   Compiling evmaptest v0.1.0 (/home/frank/[...]/evmaptest)
    Finished dev [unoptimized + debuginfo] target(s) in 0.78s
     Running `target/debug/evmaptest`
[1]    30645 segmentation fault (core dumped)  cargo run
```

<hr>

Change to `struct Sneaky(u8);` and `Box::new(Sneaky(0))` and you get
```rust
free(): double free detected in tcache 2
[1]    31123 abort (core dumped)  cargo run
```
instead.

**Comment by jonhoo** [comment]
Ouch, yeah, that's quite the bummer. I honestly don't know of a good way to work around that beyond encoding a _bunch_ of additional information in the oplog in `apply_first` that lets you repeat the exact same operation in `apply_second` without relying on the determinism of `PartialEq`. I actually worry more about `Hash` than I do about `PartialEq`, since we're going to need raw entry APIs for both `HashMap` and `HashBag` to make it work, and even that might not be enough.

I'll have to do some thinking about this... It may be that the "solution" is to make the whole thing `unsafe`, though that's obviously not great either. Hmm hmm hmm...

**Comment by jonhoo** [comment]
One option is to make `evmap::new` and friends `unsafe`. It's pretty ugly, but there may be no good way around it...

**Comment by RustyYato** [comment]
Another option is to have an unsafe marker trait, 

```rust
pub unsafe trait ConsistentHashEq: Hash + Eq {}
```

That way you can keep `new` safe, and only allow types which provide a nice `Hash + Eq` implementation. You can also provide a derive so that downstream users don't need to write `unsafe` in the common case.

**Comment by jonhoo** [comment]
@RustyYato Unfortunately that won't quite work. `Hash` and `Eq` should _already_ be deterministic. The challenge is if someone _doesn't_ follow that contract, and that's beyond your control or knowledge. The same problem would apply to `ConsistentHashEq`. It would perhaps push an issue more into the realm of the malicious, but it would be a way to circumvent safety nonetheless.

**Comment by RustyYato** [comment]
> Hash and Eq should already be deterministic. The challenge is if someone doesn't follow that contract, and that's beyond your control or knowledge.

Yes, `ConsistentHashEq` is unsafe to implement, so ot puts the burden of proof on the type to ensure that it is correctly implementing `Hash` and `Eq`. If they don't follow the contract, then they have written unsound unsafe code, which you don't have to account for.

This has the nice benefit that you can provide a completely safe interface, if you also provide a derive macro, and implementations for `std` types

Note: using unsafe traits to ensure properties of other traits is something the `std` does. See `TrustedLen` for iterators. It enables a variety of optimizations that are only valid if `size_of` is correct. If `size_of` is incorrect and you implement TrustedLen, you will get UB.

**Comment by jonhoo** [comment]
Hmm, I'm worried about saying "it's up to the implementors of the trait", since the unsafety is then still hard to discover — if someone implements `ConsistentHashEq` maliciously in their crate, and then someone else relies on that crate, they now have unsound behavior with no unsafe code in _their_ crate. Maybe that's fine, but it _feels_ weird. The precedent from the standard library is good though.

The other downside of this is that it introduces the a trait requirement on keys and values that implementors do not have control over. For example, if someone wants to have their keys or values be `bytes::Bytes`, they now need to create a newtype and derive the trait for that. It's `ShallowCopy` all over again, except _also_ for keys. A derive helps, but not a lot, since it doesn't take care of the newtype case (as the inner type would not implement `TrustedHashEq`).

Ultimately, if users have to either unsafely implement `TrustedHashEq` or the call to `evmap::new` is unsafe, I'm inclined to say that having `new` be unsafe is more ergonomic. It means you only have to check the safety in the place where you instantiate the map, and you don't need any newtypes or additional trait implementations in `evmap` itself.

**Comment by steffahn** [comment]
> if someone implements `ConsistentHashEq` maliciously in their crate, and then someone else relies on that crate, they now have unsound behavior with no unsafe code in _their_ crate. Maybe that's fine, but it _feels_ weird. The precedent from the standard library is good though.

That’s how `unsafe` code and inparticular `unsafe trait`s work. If a library author creates an `unsafe impl` that violates the contract of `ConsistentHashEq` (i.e. `Hash` or `Eq` is not deterministic) then that `unsafe` code is "unsound". It is not undefined behavior in and by itself but it allows for other, `unsafe`-free code to trigger undefined behavior. Some also call this "library-UB", as in, the `rust-evmap` library defines that non-conforming `ConsistentHashEq` implementations can lead to undefined behavior. "library-UB" in contrast to "compiler-UB" which is when some code directly violates safety contracts that the compiler imposes (e.g. dereferencing an invalid memory address or reading a `u8` from uninitialized memory).

> if someone wants to have their keys or values be `bytes::Bytes`, they now need to create a newtype and derive the trait for that. It's `ShallowCopy` all over again, except _also_ for keys.

True. That’s a general problem with new traits from not super-popular crates (like e.g. `serde`).

> A derive helps, but not a lot

I’m no expert in derive macros, but I guess a derive for `ConsistentHashEq` would need to both make sure that the same type also _derives_ `Eq` and `Ord` (if something like this is possible) and it has to ensure that all the type’s fields are `ConsistentHashEq` as well (e.g. with generic bounds on generic type parameters). This can mostly only help _users_ of evmap to safely derive `ConsistentHashEq` for their custom structs made up of standard library types (and perhaps newtypes with unsafe implementations as above). The newtypes _are_ not very ergonomic, as you said:

> Ultimately, if users have to either unsafely implement `TrustedHashEq` or the call to `evmap::new` is unsafe, I'm inclined to say that having `new` be unsafe is more ergonomic. It means you only have to check the safety in the place where you instantiate the map, and you don't need any newtypes or additional trait implementations in `evmap` itself.

But do note that the `ConsistentHashEq` solution does not preclude the "unsafe `new`" solution. One can simply offer two functions for constructing, a safe `new` with a `ConsistentHashEq` bound and using the `DefaultHasher` hasher, and another (yet to be named better) `new_asserting_consistency` function that’s unsafe and only needs `Hash + Eq`. Furthermore, the `with_hasher` construction has to be unsafe anyways, since otherwise the `Hasher` itself (or the `BuildHasher`) could be non-deterministic. (Unless we want marker traits for something like `ConsistentBuildHasher`, too.)

**Comment by jonhoo** [comment]
Yeah, I agree it meets the requirements, or maybe expectations, for how safety works in Rust. I'm more trying to think about how users may be surprised by unsoundness. And "another crate lied" and now I have a safety error in code that looks like it's completely safe (unless the user specifically thinks to look up the definition of `ConsistentHashEq`). It just seems less likely to cause surprise to me if `new` (and `with_hasher`) are `unsafe` — "huh, these are unsafe, let me check why, oh, that's easy". Not to mention it allows us to forego a new trait, implementations for a [tonne of standard library types](https://doc.rust-lang.org/std/hash/trait.Hash.html#implementors), and a derive macro.

To me, the argument for why _not_ to make `new` unsafe is that it means users do not have to write `unsafe { }` in their code, and that's more or less it. I agree that _in general_ that's a good thing to reach for, though in this case it feels like a red herring. Sure, they don't have to write `unsafe { }`, but they still have to _think_ about unsafety in the form of the implementations of `TrustedHashEq` that they rely on. And that, I think, is easier to miss.

There is perhaps an argument for "library authors are more likely to know if their operations are deterministic than the users of those libraries are", and that's valid, except that it relies on lots of library authors adding dependencies on `evmap` and adding an implementation of `TrustedHashEq`, which _seems_ unlikely to happen. Instead, pretty much all users that go outside the standard library for their types will have to use the newtype pattern.

**Comment by RustyYato** [comment]
> [...] but they still have to think about unsafety in the form of the implementations of `TrustedHashEq` that they rely on. And that, I think, is easier to miss.

No, only the implementors have to think about the implementations of `TrustedHashEq`, the users of `evmap::new` don't. 

> [...] except that it relies on lots of library authors adding dependencies on evmap and adding an implementation of `TrustedHashEq`, which seems unlikely to happen. Instead, pretty much all users that go outside the standard library for their types will have to use the newtype pattern.

That's true, a nicer solution would be to have some sort of common crate that would provide traits like these, but I don't think such a crate exists right now. Do you know if your users need to depend on types outside of what they define and `std`?

**Comment by jonhoo** [comment]
> > [...] but they still have to think about unsafety in the form of the implementations of `TrustedHashEq` that they rely on. And that, I think, is easier to miss.
> 
> No, only the implementors have to think about the implementations of `TrustedHashEq`, the users of `evmap::new` don't.

Ah, I think we've gotten to the crux of the issue. I completely agree that _technically_ that is true. That is, _technically_ the onus is on implementors of the trait to get it right. But, if you run `evmap::new` and the types you use are foreign, you really ought to go check the implementation to see that it is not malicious, otherwise you will introduce a soundness issue in your code. I suppose it's true _in general_ that you have to trust the unsafety in all of your dependency, but this one just seems particularly egregious. But maybe I'm just overly paranoid?

In some sense, another argument here is that `Hash` and `Eq` should _always_ be deterministic. It's sort of in their contract. So why require the unsafe intermediate trait at all? What does it change in terms of the trust model for whether your code is sound? It feels like all the `unsafe trait` really does is (trivially) satisfy the requirement that you cannot have unsoundness from only safe code, without _actually_ giving any more guarantees than `Hash + Eq` already _should_ give.

Which then brings me back to: why is this nicer than making `new` `unsafe`?

> > [...] except that it relies on lots of library authors adding dependencies on evmap and adding an implementation of `TrustedHashEq`, which seems unlikely to happen. Instead, pretty much all users that go outside the standard library for their types will have to use the newtype pattern.
> 
> That's true, a nicer solution would be to have some sort of common crate that would provide traits like these, but I don't think such a crate exists right now.

No, such a crate doesn't exist, and even if it did, chances are you'd never want to depend on it. It would have to add _every other crate_ as a dependency to implement the trait for those types, so it would be hugely expensive to have as a part of the build graph.

> Do you know if your users need to depend on types outside of what they define and `std`?

Hard to say, but I'd be inclined to say that this is a common use-case for a collection type. We've seen people in issues on this repository ask for support for both `bytes` and `crossbeam_channel` for example, and I'm sure those aren't the only ones. This is _especially_ true for values, which `evmap` would also need to require this impl for.

**Comment by RustyYato** [comment]
Ok, in that case making new unsafe is fine. I thought it would be nice to provide a safe interface, but if that's not tenable, that's ok.

> It would have to add every other crate as a dependency to implement the trait for those types, so it would be hugely expensive to have as a part of the build graph.

This is not the case, otherwise crates like `serde` couldn't work. But, until such a crate hits some critical mass, it wouldn't be useful. Which is an unfortunate chicken-egg problem.

**Comment by jonhoo** [comment]
> Ok, in that case making new unsafe is fine. I thought it would be nice to provide a safe interface, but if that's not tenable, that's ok.

I also think providing a safe interface would be nice, but only if it is actually safe to use, not if it just _seems_ safe to use. If that makes sense?

> > It would have to add every other crate as a dependency to implement the trait for those types, so it would be hugely expensive to have as a part of the build graph.
> 
> This is not the case, otherwise crates like `serde` couldn't work. But, until such a crate hits some critical mass, it wouldn't be useful. Which is an unfortunate chicken-egg problem.

I think maybe we're talking past each other. There is no one crate that implements `serde::{Serialize,Deserialize}` for lots of other crates. Instead, what has happened (luckily) is that almost every crate out there implement `serde::*` for their own types, which means that taking a requirement of `T: Serialize` is actually feasible. In our case, it's unlikely that most crates would pick up a dependency on `evmap` just to implement `TrustedHashEq` I think :sweat_smile:

**Comment by RustyYato** [comment]
> I also think providing a safe interface would be nice, but only if it is actually safe to use, not if it just seems safe to use. If that makes sense?

Your argument  almost sounds like a slippery slope argument, because if you  can't trust unsafe code to do the right thing, what can you trust? For a stable example, we have the `Send` and `Sync` marker traits, whose sole purposes is to say some property exists. They are unsafe to implement, so if you do it incorrectly you can indroduce unsoundness that's visible only in safe code. So, we should be able to trust the implementors of unsafe traits, otherwise we can't get anything done.

More examples of unsafe traits:
* all of `lock_api` traits
* `arrayvec`/`smallvec`s `Array` trait
* `bytemuck`'s `Pod` trait

While unsafe traits aren't common, they are used to enforce guarantees that other unsafe code require. The difference between a normal trait and an unsafe trait is for a normal trait you *should* follow the api guidelines, and for unsafe traits you ***must*** follow the api guidelines. This is enough of a distinction for unsafe code to rely on unsafe traits, but not normal traits.

> In our case, it's unlikely that most crates would pick up a dependency on evmap just to implement `TrustedHashEq` I think

Which is why it would be nice if a separate, smaller crate provides `TrustedHashEq`, then adding the dep could be easily justified for others. Like so `serde` splits it's traits from the implementations of the serializer.

But like @steffahn said, this isn't one or the other you can provide both, but if that's too much of a maintenance burden, that's ok, just an unsafe new also works.

I don't think you need to implement this for all std types, only the common ones, integers, slices, references, smart pointers, and applicable collections. Beyond that people can use unsafe new. You only need to cover the common case with the safe version.

**Comment by jonhoo** [comment]
> > I also think providing a safe interface would be nice, but only if it is actually safe to use, not if it just seems safe to use. If that makes sense?
> 
> Your argument almost sounds like a slippery slope argument, because if you can't trust unsafe code to do the right thing, what can you trust?

Yup, you're completely right, it _is_ a slippery slope argument. Or rather, it stems from a deep-seated desire to avoid making people write unsafe wherever possible, unless they absolute have to, because I know how easy it is to get wrong. The temptation to just "throw an `unsafe impl TrustedHashEq for Foo {}` in there" is quite high. And I feel like we're less likely to have users get into a bad spot if we instead say that the unsafety happens when you _use_ a type than when you produce a type. My argument isn't so much that "unsafe traits are bad", as it is "only make people use unsafe if there isn't another way". But I also acknowledge that my argument is more "it feels like the wrong trade-off to me" than "it's objectively better".

> I don't think you need to implement this for all std types, only the common ones, integers, slices, references, smart pointers, and applicable collections. Beyond that people can use unsafe new. You only need to cover the common case with the safe version.

Actually, this gave me an idea that I think I _would_ be comfortable with: to make `TrustedHashEq` (maybe under the name `KnownStableHashEq`) a sealed trait. We would _only_ implement it inside evmap for common types that we've checked ourselves, and it would enable users that want to use any of those types to use a completely safe interface without any chance of foot-shooting. And if they want to use third-party types, they'll have to use the unsafe constructor to acknowledge that they've checked the type. How does that sound?

**Comment by steffahn** [comment]
> And if they want to use third-party types, they'll have to use the unsafe constructor to acknowledge that they've checked the type. How does that sound?

I feel like the option to derive `TrustedHashEq` for user-defined structs that e.g. only consist of standard-library types has its value too. Having the trait sealed instead of `unsafe` does take this option away.

But: A sealed trait can always be upgraded into an unsafe trait later (without any breaking changes), so I guess it could be reasonable to just start with the sealed version and see how big the need for some safe derive macros really is.

**Comment by jonhoo** [comment]
A relevant point in this discussion is that I'm also planning to extract left-right out of this repository, and then write a separate implementation that's just a straight hash table (so no multi-value), which I think will make the need for custom types much less pressing. Because in that world, all we need is to make sure the key type has a sane `Hash + Eq`, and I think it's far less likely that people use complex key types.

Relatedly: not sure what'll happen to the multi-value map with that change, especially since I'm not really using it any more. I think I'll probably move it out to its own repository managed by people who are still tinkering with Noria. Stay tuned!
